scipy/spatial/ckdtree/src/partial_sort.h

```diff
@@ -1,33 +1,50 @@
 #ifndef CKDTREE_PARTIAL_SORT
 #define CKDTREE_PARTIAL_SORT
 
-/* Adapted version of the code originally 
+/* Adapted version of the code originally
  * written by @jiefangxuanyan for scikit-learn.
  *
  */
 
-
 #include <algorithm>
 
 struct IndexComparator {
+
     const double *data;
-    ckdtree_intp_t split_dim, n_dims;
+    ckdtree_intp_t split_dim;
+    ckdtree_intp_t n_dims;
+
+    IndexComparator(const double *data,
+                    ckdtree_intp_t split_dim,
+                    ckdtree_intp_t n_dims) :
+                                            data(data),
+                                            split_dim(split_dim),
+                                            n_dims(n_dims) {};
 
     inline bool operator()(ckdtree_intp_t a, ckdtree_intp_t b) {
-        return data[a * n_dims + split_dim]
-               < data[b * n_dims + split_dim];
+        if (data[a * n_dims + split_dim] == data[b * n_dims + split_dim]) {
```

Comment:
Use ```NPY_UNLIKELY``` on this comparison as we know it is unlikely to be true.

Reply:
This is called repeatedly in a performance critical loop, So the compiler need all the information it can get. Also, you can reduce the number of ```data``` lookups by reading the two data points into two local variables (declared const double).